### PR TITLE
Possibility to run with randomly distributed primary vertices

### PR DESCRIPTION
**Pull request by iouribelikov**
This PR opens the possibility to do basic performance studies on events with randomly distributed primary vertices.  The simulation, reconstruction and comparison work in both triggered and continuous readout mode.  The information about the primary vertices is passed to the tracking via a "primary vertexer".   Limitation of the moment :  this vertexer (TrivialVertexer) takes the vertices directly from the MC.

### CODE REVIEW

File: Detectors/ITSMFT/ITS/reconstruction/include/ITSReconstruction/CookedTrackerTask.h [hunk 1]
```diff
@@ -43,22 +44,23 @@ class CookedTrackerTask : public FairTask
 
   void setContinuousMode(bool mode) { mContinuousMode = mode; }
   bool getContinuousMode() { return mContinuousMode; }
-  
+
+  TrivialVertexer& getVertexer() { return mVertexer; }
+
```
Comment (shahor02) [inline Detectors/ITSMFT/ITS/reconstruction/include/ITSReconstruction/CookedTrackerTask.h]: Do we really need to hard-code the vertexer grabbing the MC-truth vertex into the CookedTrackerTask? I think the vertex (real or idealized one) should be reconstructed in a separated component and taken by the tracker as an input. Hardcoding idealized vertexer into the tracking we will get problem with using real external vertex. In any case, even if we want the ITS vertex to serve just to this reconstruction, I think it is better to do this via generic vertexer plugin, from which then the trivial vertexer  (or others) can be derived.

File: macro/run_trac_its.C [hunk 2]
```diff
@@ -11,81 +11,86 @@
...
-        cout << endl << endl;
-        cout << "Macro finished succesfully." << endl;
-
-        std::cout << endl << std::endl;
-        std::cout << "Output file is "    << outputfile.str() << std::endl;
-        //std::cout << "Parameter file is " << parFile << std::endl;
-        std::cout << "Real time " << rtime << " s, CPU time " << ctime
-                  << "s" << endl << endl;
+void run_trac_its(Int_t nEvents = 10, TString mcEngine = "TGeant3", float rate = 0.)
+{
```
Comment (shahor02) [inline macro/run_trac_its.C]: This hard-coded input file schema is not compatible with o2sim output. In the large PR https://github.com/AliceO2Group/AliceO2/pull/862 related to TPC-ITS matching I've modified this macro to take explicitly input files real names as arguments. For backward compatibility I left this old style interface (it just composes the filenames and calls new method: https://github.com/shahor02/AliceO2/blob/dbc582b88e51567269348fa444e2fe60a55e06e8/macro/run_trac_its.C, but I am afraid some conflicts will need to be solved to make it compatible with this one.

File: Detectors/ITSMFT/ITS/reconstruction/include/ITSReconstruction/CookedTrackerTask.h (hunk 1 above)
Comment (iouribelikov) [reply Detectors/ITSMFT/ITS/reconstruction/include/ITSReconstruction/CookedTrackerTask.h]: @shahor02 Sure.  A better interface is coming.

File: macro/run_trac_its.C (hunk 2 above)
Comment (iouribelikov) [reply macro/run_trac_its.C]: @shahor02  No problem. I will resolve these conflicts,  and will add to this PR a better interface to the vertexer.

**Comment by iouribelikov** [comment]
@shahor02 @sawenzel Hi Ruben, Sandro,
The conflicts have been resolved. The functionality tests pass.

But I cannot see what went wrong with the "o2checkcode". The log is not available...

Also, I would need a bit more time for switching to the new interfaces.

So, we have two possibilities:
1) We could merge this PR, so that the functionality is there.  When ready, I will open a new PR devoted to the new interfaces.
2) We could "suspend" this PR. I would then continue adding changes here, and we will merge it altogether, once ready.

Let me know what you think is the best.

**Comment by shahor02** [comment]
Hi @iouribelikov 

If you don't mind, I would prefer to avoid hardcoded dependence on the mc-truth vertexer in the tracking code. Actually, I noticed that we still have the general vertex class, will commit it. Shall I also create a base class for vertexers (the virtual inheritance of specific vertexers should be ok, given the it it used only once per RO frame).

Cheers,
 Ruben

**Comment by iouribelikov** [comment]
@shahor02 Actually, the mc-truth overexert is *not* hard-coded into the tracking code.
The vertex info (from MC, or later on from a realistic reconstruction) is passed to the tracking via the setVertices(...) function of the tracker.  

This allows us to study the tracking performance on multi-vertex time frames already now. While we are still waiting for the real vertexer to come.  That's the idea of this PR.

The MC file is opened from the driving macro, and the MC-truth "vertexer" is created inside the TrackerTask.  Both the macro and the task are, in fact, parts of the "old interface" that is anyway going to die.  I already know how to do it better.  

But unfortunately, I am now forced to switch to something else,  for about a week or so.
That's why I asked your opinion on what to do with this PR.

**Comment by shahor02** [comment]
Hi @iouribelikov 
By hardcoded I meant it cannot take other vertices (true, it is about the task and not the tracker itself).
Anyway, mine was not a strong opinion, if you will fix this later, I think it is ok to merge this PR now.
Cheers,
 Ruben

**Comment by iouribelikov** [comment]
@shahor02 Thank you, Ruben. 
@sawenzel Then, if it is OK also with you, Sandro, we may consider merging this PR.
(and I will do my best to be back to improving the situation as soon as possible.)

**Comment by sawenzel** [comment]
build failure in O2 seems unrelated (to a new python problem with matplotlib??). This was building fine yesterday. Merging.

**Pull request merged**